virtwho/virt/vdsm/stomp.py

```diff
@@ -0,0 +1,242 @@
+        if not self.headers:
+            return b''
+        return b'\n'.join(
+            [b':'.join(
+                [
+                    self._escape_bytes(key.encode('utf-8')),
+                    self._escape_bytes(unicode(value).encode('utf-8')),
+                ]
+            ) for key, value in sorted(self.headers.items())]  # sort the headers for predictable messages
+        ) + '\n'
+
+    @classmethod
+    def _decode_headers(cls, data):
+        headers = {}
+        for line in data:
+            key, value = line.split(b':')  # TODO err intelligently if more than one ':' detected
+            key = cls._unescape_bytes(key).decode('utf-8')
+            value = cls._unescape_bytes(value).decode('utf-8')
+            if key not in headers:  # first value wins
+                headers[key] = value
+        return headers
+
+    @staticmethod
+    def _escape_bytes(data):
+        if data is None:
+            return b''
+        data = data.replace(b'\\', b'\\\\')  # \ -> \\
+        data = data.replace(b'\r', b'\\r')   # CR -> \r
+        data = data.replace(b'\n', b'\\n')   # LF -> \n
+        data = data.replace(b':', b'\\c')    # : -> \c
+        return data
+
+    @staticmethod
+    def _unescape_bytes(data):
+        data = data.replace(b'\\c', b':')    # \c -> :
+        data = data.replace(b'\\n', b'\n')   # \n -> LF
+        data = data.replace(b'\\r', b'\r')   # \r -> CR
+        # TODO could check here that there are illegal escape sequences (ex. \t), see STOMP spec for details
+        data = data.replace(b'\\\\', b'\\')  # \\ -> \
+        return data
+
+
+class StompClient(object):
+    """Minimal STOMP client with SSL support."""
+
+    def __init__(self, host, port, ssl_context=None, timeout=None):
+        """Create a new STOMP client.
+
+        :param host: STOMP server hostname.
+        :param port: STOMP server port.
+        :param ssl_context: SSL context if using TLS.
+        :param timeout: timeout for blocking operations.
+        """
+        self.host = host
+        self.port = port
+        self.ssl_context = ssl_context
+        self.timeout = timeout
+        self.socket = None
+        self.buffer = b''  # FIXME for performance, could construct a buffer and then recv_into instead...
+        self.subscription_ids = []
+
+    def __enter__(self):
+        self.connect()
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.close()
+
+    def connect(self):
+        """Connect to the STOMP server.
+
+        Sends a STOMP CONNECT and consumes the subsequent CONNECTED message.
+
+        :raises: IOError if no connection attempts succeed.
+        """
+        for family, _, _, _, addr in socket.getaddrinfo(self.host, self.port):
+            try:
+                self.socket = socket.socket(family)
+                self.socket.settimeout(self.timeout)
+                if self.ssl_context:
+                    self.socket = self.ssl_context.wrap_socket(self.socket)
```

Comment:
This raises error on RHEL 7, because `self.ssl_context.wrap_socket` is not available here :-(

Reply:
Had logic in for this, and removed unintentionally. I added it back in latest commit.